km/km_filesys.c

```diff
@@ -1131,7 +1131,7 @@ uint64_t km_fs_dup3(km_vcpu_t* vcpu, int fd, int newfd, int flags)
    assert(name != NULL);
    ret = __syscall_3(SYS_dup3, host_fd, newfd, flags);
    if (ret >= 0) {
-      if (km_fs()->guest_files[ret].inuse != 0) {
+      if (km_is_file_used(&km_fs()->guest_files[ret]) != 0) {
          del_guest_fd(vcpu, ret);
       }
       ret = km_add_guest_fd(vcpu, ret, name, flags, ops);
```

Comment:
It feels like this entire sequence: dup3(), is_file_used test, del_guest_fd(), add_guest_fd() should be an atomic operation.

If you have 2 (or more) threads dup3()ing onto the same newfd, the results would vary depending on whether the involved threads are preempted or not.  We would probably see assert failures in del_guest_fd() or km_add_guest_fd_internal().

Reply:
Yes this is totally valid, but here is what I'm going to do:

Going to merge without dup3 changes suggested here. Opened Issue #1479 to capture this.